File: c/src/bead_sort.c
#include "../sorting_algorithms.h"
#include <stdlib.h>
#include <string.h>

void bead_sort(int *arr, size_t n) {
    if (n == 0) return;
    int min = arr[0], max = arr[0];
    for (size_t i = 1; i < n; ++i) {
        if (arr[i] < min) min = arr[i];
        if (arr[i] > max) max = arr[i];
    }
    if (min < 0) { counting_sort(arr, n); return; }
    unsigned char *grid = (unsigned char *)calloc(n * max, sizeof(unsigned char));
    if (!grid) return;
    for (size_t i = 0; i < n; ++i) {
        for (int j = 0; j < arr[i]; ++j) grid[i * max + j] = 1;
    }
    for (int j = 0; j < max; ++j) {
        size_t sum = 0;
        for (size_t i = 0; i < n; ++i) sum += grid[i * max + j];
        for (size_t i = n; i-- > 0;) {
            grid[i * max + j] = (sum > 0);
            if (sum > 0) sum--;
        }
    }
    for (size_t i = 0; i < n; ++i) {
        int j;
        for (j = 0; j < max && grid[i * max + j]; ++j);
        arr[i] = j;
    }
    free(grid);
}

File: python/sorting_algorithms/algos/bead_sort.py
from typing import Sequence, Optional, List
from ..base import _setup_tracer, _finalize, SortTracer, Number


def bead_sort(data: Sequence[int], *, trace: bool = False, tracer: Optional[SortTracer] = None):
    arr = list(data)
    tracer, ret_trace = _setup_tracer("bead_sort", trace, tracer)
    if not arr:
        return _finalize(arr, tracer, ret_trace)

    min_val = min(arr)
    if min_val < 0:
        # shift to non-negative domain; bead sort assumes >=0
        shift = -min_val
        arr = [x + shift for x in arr]
    else:
        shift = 0

    max_val = max(arr)
    grid = [[0] * len(arr) for _ in range(max_val)]

    for col, val in enumerate(arr):
        for row in range(val):
            grid[row][col] = 1
    if tracer:
        tracer.record(arr, (), "load beads")

    for row in grid:
        beads = sum(row)
        row[:] = [1] * beads + [0] * (len(arr) - beads)

    output: List[int] = []
    for col in range(len(arr)):
        count = sum(grid[row][col] for row in range(max_val))
        output.append(count)
    output.reverse()  # natural bead drop yields descending; reverse to ascending
    if shift:
        output = [x - shift for x in output]
    if tracer:
        tracer.record(output, (), "collect")
    return _finalize(output, tracer, ret_trace)
